src/rules/arrowParens.js

```diff
@@ -0,0 +1,169 @@
+const getLocation = (node) => {
+  return {
+    end: node.params[node.params.length - 1].loc.end,
+    start: node.params[0].loc.start
+  };
+};
+
+const isOpeningParenToken = (token) => {
+  return token.value === '(' && token.type === 'Punctuator';
+};
+
+const isClosingParenToken = (token) => {
+  return token.value === ')' && token.type === 'Punctuator';
+};
+
+export default {
+  create (context) {
+    const asNeeded = context.options[0] === 'as-needed';
+    const requireForBlockBody = asNeeded && context.options[1] && context.options[1].requireForBlockBody === true;
+
+    const sourceCode = context.getSourceCode();
+
+    // Determines whether a arrow function argument end with `)`
+    // eslint-disable-next-line complexity
+    const parens = (node) => {
+      const isAsync = node.async;
+      const firstTokenOfParam = sourceCode.getFirstToken(node, isAsync ? 1 : 0);
+
+      // Remove the parenthesis around a parameter
+      const fixParamsWithParenthesis = (fixer) => {
+        const paramToken = sourceCode.getTokenAfter(firstTokenOfParam);
+
+        /*
+         * ES8 allows Trailing commas in function parameter lists and calls
+         * https://github.com/eslint/eslint/issues/8834
+         */
+        const closingParenToken = sourceCode.getTokenAfter(paramToken, isClosingParenToken);
+        const asyncToken = isAsync ? sourceCode.getTokenBefore(firstTokenOfParam) : null;
+        const shouldAddSpaceForAsync = asyncToken && asyncToken.range[1] === firstTokenOfParam.range[0];
+
+        return fixer.replaceTextRange([
+          firstTokenOfParam.range[0],
+          closingParenToken.range[1]
+        ], `${shouldAddSpaceForAsync ? ' ' : ''}${paramToken.value}`);
+      };
+
+      // Type parameters without an opening paren is always a parse error, and
+      // can therefore be safely ignored.
+      if (node.typeParameters) {
+        return;
+      }
```

Comment:
This is the relevant bit. My assumption is that if the arrow function has type parameters, then we can safely ignore it because any arrow function with type parameters and no parens is always a parse error.